### ISSUE TITLE
[Enhancement]: Finished podcast episode handling

### ISSUE THREAD

**Issue by andonevris** [advplyr/audiobookshelf#615]
### Describe the feature/enhancement

I think we should have the ability to handle what happens once you have finished listening to a podcast episode, namely deleting media files either immediately or via a size quota.

Some podcasts are transient by nature so some media file management features would be very useful to clear out old files.

**Comment by advplyr** [comment]
I agree. I think the same needs to be done with the mobile app but for locally downloaded episodes.

I'm hoping someone will volunteer to do a mockup of how this should all look and function. That is usually the hardest part for me, building it is the fun part.

**Comment by andonevris** [comment]
I use antennapod for podcasts (open source) they handle finished podcasts well.

Global settings for storage:
Toggle switch to turn auto delete on off
Toggle switch to keep favourite episodes (need to be able to mark episode as favourite)
Toggle switch "Delete removes from queue" (once you build out the queueing of episodes it's useful to remove deleted episodes from that too)

Each podcast has its own settings: 
Auto delete episodes  checkbox with -  "global default" : "always" : "never"

**Comment by shepda** [comment]
Perhaps this feature needs to take multiple users into consideration because if a finished episode is deleted it may be deleted before other users have listened to it?

I think it would be nice if a finished episode is hidden from that user's Newest Episodes list.

**Comment by advplyr** [comment]
Good point. I think this feature makes more sense on mobile devices then it does on your media folder.

I've already started adding a filter in the mobile app so you can default to filtering out finished episodes which I will also add to the web app.

**Comment by advplyr** [comment]
Most podcast players are mobile apps or don't have multi-user.
I'm thinking what we will do is the mobile apps will have settings for auto-deleting episodes. The server will have the filter to hide finished episodes and have the option to select multiple episodes and batch delete.

**Comment by Lebo77** [comment]
An ability of what order to show the podcast episodes by default would be smart as well. Some you want to go oldest-first. Others you want to see the most recent first. 

An optional limit on the number of episodes, and which ones get deleted and when would be a good addition.

**Comment by advplyr** [comment]
I was thinking of adding this for the next release then realized this will be problematic for a server with many users.
This might only make sense to do on mobile and remove the episode if it is downloaded on mobile.

**Comment by FuzzyMistborn** [comment]
On mobile would be a great start. For the server that's a good point on multi user servers. Could it be added as a setting that could be enabled manually?

**Comment by advplyr** [comment]
A setting that does what though? Which user would be the one that determines whether the item is finished and ready to remove?

**Comment by FuzzyMistborn** [comment]
I guess there's no admin account?

**Comment by advplyr** [comment]
There can be many admin accounts

**Comment by andonevris** [comment]
In AB could we possibly make the podcast and the users subscribed to that podcast as separate entries?

For instance I add a podcast and only I am subscribed to it within AB. If other users want to listen to that podcast it already exists in AB so they can subscribe to it also.

This way we can have settings like "delete episode after last subscriber has listened" 

IMO it's very unlikely all users will be interested in the same podcast, in my household there is no overlap in our podcast tastes at all, this way we can delete files once anyone interested has finished listening.

**Comment by advplyr** [comment]
That seems very complicated in the current system of libraries and user permissions. Maybe, but not something I can picture right now.

**Comment by RandomWare** [comment]
I second the filter option. I sometimes have a backlog of episodes built up. When I want to catch up I would ideally sort by release date and start playing the first episode in the list, which would be the oldest. When that episode is finished, the player would queue up the next one, and so on until there are no more unread episodes. 

I am also a data hoarder, so would prefer not to delete old episodes, even if I've already listened to them. I was using podgrab until recently to automatically maintain a backup of my podcasts, but I've effectively replaced it with this app. However, since I don't want to delete the files, but also don't want to have them show up in the list of episodes when sorting by oldest first, it would be best to have a way to hide episodes that are already finished.

**Comment by gbakeman** [comment]
I'd just like to add to this issue: I use ABS primarily as my personal podcast management system. I'd like it so that episode listen state is not tied to the episode file existing on the server, so I can clean up old episodes but still retain the metadata and listen state. Hoping this can be considered.

**Comment by advplyr** [comment]
@gbakeman I'm not sure what you mean. Your listening progress if you are using the abs mobile app or the web client will always be tied to the episode on the server.

**Comment by gbakeman** [comment]
The issue for me though is, when we "hard delete" a listened episode (only option to remove the actual file if I understand correctly), then any evidence that that episode was listened to is also gone (can't tell from the episodes list window anyways).

I think it's basically just how I'm used to most RSS reading apps - typically when you add a new feed, all of the metadata (or the last _x_ entries) are downloaded and viewable, then you mark as read from the main view. Of course this is a different situation when dealing with actual files that you need to download, but my idea is that all (or x amount of) episodes in a given feed are viewable directly on the main interface, instead of needing to open the additional episode list window. You'd have various filters to make the list more manageable i.e hide listened, show only downloaded, etc. but otherwise the feed's list retains your watch status regardless of whether or not the file is stored on the server.

Hope I'm making sense and not trying to totally change up how ABS works. Thank you for your work!

**Comment by advplyr** [comment]
It makes sense. I'm thinking we will remove the search for episode modal and just put the RSS feed in the same spot downloads are. Then the default view can be filtered for what you have downloaded but you can remove that filter and see the full RSS feed of episodes with download buttons.

**Comment by gbakeman** [comment]
> It makes sense. I'm thinking we will remove the search for episode modal and just put the RSS feed in the same spot downloads are. Then the default view can be filtered for what you have downloaded but you can remove that filter and see the full RSS feed of episodes with download buttons.

That sounds ideal. Do you want a separate issue opened for this or do you just want to track that here?

**Comment by advplyr** [comment]
@gbakeman A new issue for this is a good idea. Thanks

**Comment by shepda** [comment]
> I've already started adding a filter in the mobile app so you can default to filtering out finished episodes which I will also add to the web app.

I can see that this has been implemented if you click on the podcast on the library screen. I think it should also be added to the Newest episodes row on the Home Screen. Personally I use this row the most to listen to different podcast episodes in the order they were released.

**Comment by mcmikemn** [comment]
Some use cases (like mine) are single-user, but even some cases with multiple users possibly want to auto-manage media. Does anyone actually keep all media _forever_? Why not add an _option_ to groom media. If you want to keep all media, turn grooming off; if you want to delete media "after X Y has passed", turn it on (X could be a number and Y could be a length of time, e.g., X=6, Y=months - and you might even include a sub-option "groom even if not listened to?").

**Comment by alfureu** [comment]
I would like to chime into this discussion as well. Just facing the same thing. There are podcasts that I do not want to keep, other podcasts I would love to keep. 

I noticed that there are icons on the 3 corners of the cover in ABS, maybe the 4th corner can be used to trigger the keep option?

![image](https://github.com/advplyr/audiobookshelf/assets/10846169/51a932b3-458a-42f6-afa2-ff73eed963b1)

**Comment by MrSimmo** [comment]
Hi all,

I also would really like this feature. In the meantime, I have written a python script that can look for finished items and then delete them. You can also add a tag to the show to tell the script to ignore it (i.e. not delete it).

https://github.com/MrSimmo/audiobookself_auto_purge

It has the ability to filter on media type, age of media, ignore specific shows/titles  etc.

**Comment by evanpeterjones** [comment]
Chiming in with my thoughts and suggestions based on a different use case and solution.

Essentially I like the existing "Max # of episodes to keep" feature, and think it should be expanded with a Global Default and "Favorites" feature to circumvent auto-deletion.

Whether you or any user has played an episode should be irrelevant as to whether it's deleted if it is above this max.

I use Audiobookshelf both to archive limited-series fiction podcasts (ever since Homecoming went defunct and is now gone forever), as well as listen to plenty of weekly topical podcasts that I have no interest in archiving Gigabytes of episodes of forever.


My suggestion/ideal solution would be to add a global default for the existing, "Max # of episodes to keep" option and then to just add an override for this deletion process--there are two options for this that I see...
1. Add a "Favorite" at the Podcast and Episode level - if an episode is favorited, exclude this from deletion, if a podcast is favorited, never delete an episode regardless of global Max.
2. Add a manual override per-podcast for "Max # of episodes to keep" so you could set this to 0 for podcasts you want to archive. 

I prefer the 1st option because it provides more flexibility to save particular episodes, but I can understand how implementing a new Favorites feature might be more of a haul in terms of development.

**Comment by sjhuda** [comment]
I agree with @evanpeterjones! I'd love to see a "keep forever" option! I find a lot of podcasts I listen to now only have the latest 3 episodes or so and the rest of the archive is only available on their app (New York Times usually do this), so have to go rummaging through on there. Would be so much nicer to have it stored on my own server!
